File: solutions/elixir/luhn/lib/luhn.ex
defmodule Luhn do
  @doc """
  Checks if the given number is valid via the luhn formula
  """
  @spec valid?(String.t()) :: boolean
  def valid?(number) do
    number = Regex.replace(~r/[\s]/, number, "")

    cond do
      Regex.match?(~r/[^0-9]/, number) -> false
      String.length(number) < 2 -> false
      rem(luhn_sum(number), 10) == 0 -> true
      true -> false
    end
  end

  defp luhn_sum(xs) do
    xs
    |> String.graphemes()
    |> Enum.reverse()
    |> Enum.map(&String.to_integer/1)
    |> Enum.with_index()
    |> Enum.map(&modify_digit/1)
    |> Enum.sum()
  end

  defp modify_digit({digit, index}) when rem(index, 2) == 1 do
    if digit >= 5, do: 2 * digit - 9, else: 2 * digit
  end

  defp modify_digit({digit, _}), do: digit
end

File: solutions/fsharp/luhn/Luhn.fs
module Luhn

open System.Text.RegularExpressions

let (|InvalidCharacter|) (n: string) = Regex.IsMatch(n, "[^0-9]")
let (|TooShort|) (n: string) = n.Length <= 1

let (|InvalidChecksum|) (n: string) =
    n
    |> fun n -> Regex.Replace(n, " ", "")
    |> Seq.rev
    |> Seq.map (string >> int)
    |> Seq.indexed
    |> Seq.sumBy
        (function
        | i, n when i % 2 = 0 -> n
        | _, n when n >= 5 -> 2 * n - 9
        | _, n -> 2 * n)
    |> (fun n -> n % 10)
    |> (<>) 0

let valid (number: string) : bool =

    number
    |> fun n -> Regex.Replace(n, " ", "")
    |> function
        | InvalidCharacter true
        | TooShort true
        | InvalidChecksum true -> false
        | _ -> true
